### PR TITLE
Add TLS settings to tracer

### LINKED ISSUE #22741: Support TLS setting for tracer Zipkin
**Describe the feature request**
In some scenarios, users should set up TLS to encrypt the communication between isito-proxy and zipkin-server. If users use a self-signed CA, there's no chance to set it to cluster's tlscontext of Envoy. 

I would like to contribute this feature, and my current plan is to follow lightstep to add https://github.com/istio/istio/blob/master/manifests/global.yaml#L250-L251 to zipkin config. 

**Describe alternatives you've considered**

Another option is to copy https://github.com/istio/istio/blob/master/manifests/global.yaml#L185-L192, which looks like more functional than lightstep.

**Additional context**

### PR DESCRIPTION
**Pull request by hanahmily**
Signed-off-by: Gao Hongtao <hanahmily@gmail.com>

These changes intend to fix #22741. Add a generic TLS setting to most of the tracers. 

Notice: This PR is in WIP
 - [x] API change https://github.com/istio/api/pull/1397
 - [x] Dev
 - [ ] Test
 - [ ] Document

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -13,6 +13,7 @@ replace github.com/docker/docker => github.com/docker/engine v1.4.2-0.2019101121
 
 // Old version had no license
 replace github.com/chzyer/logex => github.com/chzyer/logex v1.1.11-0.20170329064859-445be9e134b2
+replace istio.io/api => github.com/hanahmily/api v0.0.0-20200422080657-e387a4b4939a
```
Comment (hanahmily) [inline go.mod:16]: Once https://github.com/tetrateio/api/pull/28/ is merged, delete this line.

**Review by hzxuzhonghu — COMMENTED** [review]
/ok-to-test

**Comment by nrjpoddar** [comment]
@hanahmily I approved your API changes, please get the implementation in 1.6 as API in not hidden.

@jacob-delgado @brian-avery @irisdingbj please track this PR to ensure it gets into release-1.6 branch if we branch before this collapses. Thanks!

File: tools/packaging/common/envoy_bootstrap_v2.json [hunk 2]
```diff
@@ -405,6 +405,9 @@
       ,
       {
         "name": "zipkin",
+        {{ if .tracing_tls }}
+        "tls_context": {{ .tracing_tls }},
```
Comment (howardjohn) [inline tools/packaging/common/envoy_bootstrap_v2.json:409]: tls_context is deprecated, we should use transport_socket

File: tools/packaging/common/envoy_bootstrap_v2.json [hunk 3]
```diff
@@ -427,21 +430,10 @@
...
-              "trusted_ca": {
-                "filename": "{{ .lightstepCacertPath }}"
-              }
-            }
-          }
-        },
+        {{ if .tracing_tls }}
+        "tls_context": {{ .tracing_tls }},
         {{ end }}
+        "http2_protocol_options": {},
```
Comment (howardjohn) [inline tools/packaging/common/envoy_bootstrap_v2.json:436]: Why is this added? It seems unrelated maybe?

Comment (kyessenov) [reply tools/packaging/common/envoy_bootstrap_v2.json:436]: This will break http/1.1 servers. There is no negotiation at client level.

**Review by kyessenov — APPROVED** [review]

File: manifests/charts/global.yaml [hunk 4]
```diff
@@ -264,6 +259,14 @@ global:
...
       # The global default max number of message events per span.
       maxNumberOfMessageEvents: 200
+    tlsSettings:
+      mode: DISABLE # DISABLE, SIMPLE, MUTUAL, ISTIO_MUTUAL
+      clientCertificate: # example: /etc/istio/tracer/cert-chain.pem
+      privateKey:        # example: /etc/istio/tracer/key.pem
+      caCertificates:    # example: /etc/istio/tracer/root-cert.pem
+      sni:               # example: tracer.somedomain
+      subjectAltNames: []
+      # - tracer.somedomain
```
Comment (nrjpoddar) [inline manifests/charts/global.yaml:269]: We shouldn't be adding these here afaik, we should move on to using Operator APIs @howardjohn correct me if I'm wrong.

**Review by nrjpoddar — CHANGES_REQUESTED** [review]
This PR is bloated, keep the changes to a minimum/what's needed.

File: manifests/charts/global.yaml (hunk 4 above)
Comment (howardjohn) [reply manifests/charts/global.yaml:269]: Yes you can directly set mesh config now. So this would be 
```
meshConfig:
  tracer:
    tlsSettings: ...
```

**Comment by nrjpoddar** [comment]
There's no progress on this PR but the istio.io/api PR https://github.com/istio/api/pull/1397 was merged which was not backward compatible as it removed fields.

Now we cannot update istio.io/istio to latest commit from istio.io/api

@howardjohn @rshriram what's the best thing to do here? Should we revert https://github.com/istio/api/pull/1397?

File: tools/packaging/common/envoy_bootstrap_v2.json (hunk 3 above)
Comment (rshriram) [reply tools/packaging/common/envoy_bootstrap_v2.json:436]: This already exists for the lightstep block. So this is not a new change

File: tools/packaging/common/envoy_bootstrap_v2.json (hunk 2 above)
Comment (howardjohn) [reply tools/packaging/common/envoy_bootstrap_v2.json:409]: can be followup

File: go.mod (hunk 1 above)
Comment (howardjohn) [reply go.mod:16]: LGTM once this changes

**Review by nrjpoddar — APPROVED** [review]
@hanahmily I read the diff wrong, please continue with this PR and fetch the latest istio.io/api. @rshriram clarified few things for me, sorry about the confusion!

**Comment by nrjpoddar** [comment]
@hanahmily can we get this merged? We would also need to cherry-pick this to release-1.6 as this is blocking my PR https://github.com/istio/istio/pull/23292 which also needs to be cherry-picked. Thanks!

**Comment by hanahmily** [comment]
@nrjpoddar I'm on it now. 

> @hanahmily can we get this merged? We would also need to cherry-pick this to release-1.6 as this is blocking my PR #23292 which also needs to be cherry-picked. Thanks!

@nrjpoddar sorry for blocking your PR, after removing some items from API, I found a number of test cases failed, so I take more time to fix up them. Most of them have passed on my local machine, I will push them soon

File: tools/packaging/common/envoy_bootstrap_v2.json (hunk 2 above)
Comment (hanahmily) [reply tools/packaging/common/envoy_bootstrap_v2.json]: done

**Comment by hanahmily** [comment]
@nrjpoddar @rshriram all test cases have passed, please help me review it, thanks.

File: manifests/charts/global.yaml [hunk 5]
```diff
@@ -568,3 +562,13 @@ version: ""
 
 meshConfig:
   enablePrometheusMerge: false
+  defaultConfig:
+    tracing:
+      tlsSettings:
+        mode: DISABLE # DISABLE, SIMPLE, MUTUAL, ISTIO_MUTUAL
```
Comment (nrjpoddar) [inline manifests/charts/global.yaml:568]: @rshriram @hanahmily can we support all 4 modes for this? I think only DISABLE/SIMPLE might actually work.

File: manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml [hunk 6]
```diff
@@ -635,8 +658,8 @@ data:
         {{- end }}
         - name: istio-podinfo
           mountPath: /etc/istio/pod
-        {{- if and (eq .Values.global.proxy.tracer "lightstep") .Values.global.tracer.lightstep.cacertPath }}
-        - mountPath: {{ directory .ProxyConfig.GetTracing.GetLightstep.GetCacertPath }}
+         {{- if and (eq .Values.global.proxy.tracer "lightstep") .Values.meshConfig.defaultConfig.tracing.tlsSettings }}
```
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml:661]: nit: Values. meshConfig.defaultConfig -> . ProxyConfig

they represent the same thing but the latter takes into account per pod annotation

File: manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml [hunk 7]
```diff
@@ -698,7 +721,7 @@ data:
         {{ toYaml $value | indent 2 }}
         {{ end }}
         {{ end }}
-      {{- if and (eq .Values.global.proxy.tracer "lightstep") .Values.global.tracer.lightstep.cacertPath }}
+      {{- if and (eq .Values.global.proxy.tracer "lightstep") .Values.meshConfig.defaultConfig.tracing.tlsSettings }}
```
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/files/gen-istio.yaml:724]: same comment about ProxyConfig

File: manifests/charts/istio-control/istio-discovery/templates/istiod-injector-configmap.yaml [hunk 8]
```diff
@@ -10,7 +10,7 @@ metadata:
 data:
 {{/* Scope the values to just top level fields used in the template, to reduce the size. */}}
   values: |-
-{{ pick .Values "global" "istio_cni" "sidecarInjectorWebhook" "revision" | toPrettyJson | indent 4 }}
+{{ pick .Values "global" "istio_cni" "sidecarInjectorWebhook" "revision" "meshConfig" | toPrettyJson | indent 4 }}
```
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/templates/istiod-injector-configmap.yaml:13]: this can be removed based on prior suggestion

Comment (nrjpoddar) [inline manifests/charts/istio-control/istio-discovery/templates/istiod-injector-configmap.yaml:13]: Why is this change needed?

**Review by howardjohn — APPROVED** [review]
this is great thanks! We can follow-up the minor comments in another PR if needed

**Review by nrjpoddar — COMMENTED** [review]
This is pretty great! Who knew, a rather simple change can lead to so many changes. Let's get this merged, if there's any minor feedback you can address in a follow-up.

File: manifests/charts/istio-control/istio-discovery/templates/istiod-injector-configmap.yaml (hunk 8 above)
Comment (howardjohn) [reply manifests/charts/istio-control/istio-discovery/templates/istiod-injector-configmap.yaml:13]: it's needed because they reference mesh config from the injection template. the injector passes them directly so we shouldn't do this, in a follow-up we can change this to unblock others

**Comment by nrjpoddar** [comment]
Please cherry-pick this release-1.6, thanks!

**Comment by nrjpoddar** [comment]
/cherrypick/release-1.6

File: manifests/charts/global.yaml (hunk 5 above)
Comment (rshriram) [reply manifests/charts/global.yaml:568]: For people running istio proxy next to zipkin, istio mutual helps.

**Review by rshriram — COMMENTED** [review]
.

**Comment by jacob-delgado** [comment]
> this is great thanks! We can follow-up the minor comments in another PR if needed

cc @hanahmily When you do this we should probably put a release-1.6 cherry pick label.

**Comment by nrjpoddar** [comment]
> /cherrypick release-1.6

I just did this manually, so no longer needed. Thanks though!

**Comment by howardjohn** [comment]
@hanahmily let me know if you are going to do the follows, if not I can work on it

**Comment by hanahmily** [comment]
> @hanahmily let me know if you are going to do the follows, if not I can work on it

@howardjohn Sorry I'm getting distracted to something else on our team project. It's great if you can help with it. Thanks a lot.
